### PR TITLE
feat: Add state-based configuration support to MCPToolset

### PR DESCRIPTION
**Pull request by vblagoje**
## Why

Agents need tools that seamlessly share data through state. `MCPTool` already supports `inputs_from_state` and `outputs_to_state` for automatic parameter injection and output handling. `MCPToolset` couldn't use these features - breaking Agent workflows that rely on state propagation across multiple MCP tools.

This PR completes MCP's Agent integration story. `MCPToolset` now dynamically configures state mappings for each discovered tool, enabling the same seamless state-based coordination that `MCPTool` provides.

- fixes https://github.com/deepset-ai/haystack-core-integrations/issues/2677

## What

Added per-tool state configuration to `MCPToolset`:
- Added `inputs_from_state`, `outputs_to_state`, `outputs_to_string` parameters (dict mapping tool names to configs)
- `_serialize_state_config()` / `_deserialize_state_config()` helpers handle both nested (outputs_to_state) and flat (outputs_to_string) formats
- `_validate_state_configs()` warns on unknown tool names
- ValueError from Tool parameter validation bubbles up (Haystack >=2.22.0)
- Updated dependency to `haystack-ai>=2.19.0` (supports state params)

## How can it be used

```python
from haystack_integrations.tools.mcp import MCPToolset, StdioServerInfo

# Create toolset with per-tool state configuration
toolset = MCPToolset(
    server_info=StdioServerInfo(command="uvx", args=["mcp-server-git"]),
    tool_names=["git_status", "git_diff", "git_log"],
    
    # Map state keys to tool parameters
    inputs_from_state={
        "git_status": {"repository": "repo_path"},
        "git_diff": {"repository": "repo_path"},
        "git_log": {"repository": "repo_path"},
    },
    
    # Map tool outputs to state keys
    outputs_to_state={
        "git_status": {"status_result": {"source": "status"}},
        "git_diff": {"diff_result": {}},
    },
    
    # Convert outputs to strings for LLM
    outputs_to_string={
        "git_diff": {"source": "diff", "handler": format_diff},
    },
)
```

## How did you test it

- 3 unit tests: state config creation, serialization roundtrip, unknown tool warning
- 1 validation test: ValueError on invalid parameter (skipped on Haystack <2.22.0)
- 6 helper tests: serialize/deserialize for both outputs_to_state and outputs_to_string formats

## Notes for the reviewer

Parameter validation behavior:
- Haystack >=2.22.0: ValueError raised at tool creation time for invalid parameter names
- Haystack 2.19.0-2.21.x: Invalid parameters fail at runtime during tool invocation

### CODE REVIEW

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 1]
```diff
@@ -99,6 +178,30 @@ class MCPToolset(Toolset):
...
+        server_info=StdioServerInfo(command="uvx", args=["mcp-server-git"]),
+        tool_names=["git_status", "git_diff", "git_log"],
+
+        # Map state keys to tool parameters for each tool
+        inputs_from_state={
+            "git_status": {"repository": "repo_path"},
+            "git_diff": {"repository": "repo_path"},
+            "git_log": {"repository": "repo_path"},
+        },
+        # Map tool outputs to state keys
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]:
```suggestion
        # Map tool outputs to state keys for each tool
```

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 2]
```diff
@@ -99,6 +178,30 @@ class MCPToolset(Toolset):
...
+        outputs_to_state={
+            "git_status": {"status_result": {"source": "status"}},
+            "git_diff": {"diff_result": {}},
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]: Could we add a dev comment here explaining why one might provide an empty dict here?

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 3]
```diff
@@ -132,14 +238,35 @@ def __init__(
         :param invocation_timeout: Default timeout in seconds for tool invocations
         :param eager_connect: If True, connect to server and load tools during initialization.
                              If False (default), defer connection to warm_up.
+        :param inputs_from_state: Optional dictionary mapping tool names to their inputs_from_state config.
+                                 Each config maps state keys to tool parameter names.
+                                 Tool names should match available tools from the server; a warning is logged for
+                                 unknown tools. Note: With Haystack >= 2.22.0, parameter names are validated;
+                                 ValueError is raised for invalid parameters. With earlier versions, invalid
+                                 parameters fail at runtime.
+                                 Example: `{"git_status": {"repository": "repo_path"}}`
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:252]: Just to clarify, what happens if eager_connect is False, is validation skipped or does it occur when the connection is made?

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 4]
```diff
@@ -99,6 +178,30 @@ class MCPToolset(Toolset):
...
+    toolset = MCPToolset(
+        server_info=StdioServerInfo(command="uvx", args=["mcp-server-git"]),
+        tool_names=["git_status", "git_diff", "git_log"],
+
+        # Map state keys to tool parameters for each tool
+        inputs_from_state={
+            "git_status": {"repository": "repo_path"},
+            "git_diff": {"repository": "repo_path"},
+            "git_log": {"repository": "repo_path"},
+        },
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]:
```suggestion
        # Maps the state key "repository" to the tool parameter "repo_path" for each tool
        inputs_from_state={
            "git_status": {"repository": "repo_path"},
            "git_diff": {"repository": "repo_path"},
            "git_log": {"repository": "repo_path"},
        },
```

Comment (sjrl) [reply integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]: Let's make the relationship super clear in the example of what is what

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 5]
```diff
@@ -292,6 +425,31 @@ def invoke_tool(**kwargs: Any) -> Any:
...
             raise MCPConnectionError(message=message, server_info=self.server_info, operation="initialize") from e
 
+    def _validate_state_configs(self, available_tool_names: set[str]) -> None:
+        """
+        Validate that state configuration tool names exist in the toolset.
+
+        Logs a warning for any tool names in the state configs that don't match
+        available tools in the toolset.
+
+        :param available_tool_names: Set of tool names that are available in the toolset
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:464]: I think it's probably fine as a warning, but just curious your motivation for making this a warning and not an error?

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 6]
```diff
@@ -324,13 +485,21 @@ def from_dict(cls, data: dict[str, Any]) -> "MCPToolset":
         server_info_class = import_class_by_name(server_info_dict["type"])
         server_info = cast(MCPServerInfo, server_info_class).from_dict(server_info_dict)
 
+        # Deserialize state configuration parameters
+        inputs_from_state = inner_data.get("inputs_from_state") or {}
```
Comment (sjrl) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]: Technically this doesn't need to be a dict by default right?
```suggestion
        inputs_from_state = inner_data.get("inputs_from_state")
```

Comment (sjrl) [reply integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py]: Not a big deal so feel free to ignore

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py (hunk 3 above)
Comment (vblagoje) [reply integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:252]: Yes, validation is deferred until warm_up() is called it is not skipped.

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py (hunk 5 above)
Comment (vblagoje) [reply integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:464]: It makes sense to me to have a warning for unknown tool names in the state config as it might be available dynamically or somehow differently. Raising ValueError makes more sense for valid parameter names within a valid tool as that's config error and should be flagged. For the former it it is debatable. Hence being lenient makes more sense imho :-)

Comment (sjrl) [reply integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:464]: Thanks for the explanation!

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 7]
```diff
@@ -27,6 +28,84 @@
...
+    """
+    Serialize a state configuration dictionary, converting any callable handlers to their string representation.
+
+    Works for both outputs_to_state (tool_name -> {state_key -> {source, handler}})
+    and outputs_to_string (tool_name -> {source, handler}).
+
+    :param config: The state configuration dictionary to serialize
+    :returns: The serialized configuration dictionary, or None if empty
+    """
+    if not config:
```
Comment (mpangrazzi) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:46]: More a type checker thing: the function accepts and handles `None` correctly, but the type hint says it only accepts `dict`.

File: integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py [hunk 8]
```diff
@@ -27,6 +28,84 @@
...
+        return None
+
+    serialized = {}
+    for tool_name, tool_config in config.items():
+        if not tool_config:
+            continue
+
+        # Check if this is outputs_to_string format (flat with optional source/handler)
+        # or outputs_to_state format (nested with state keys)
+        if "source" in tool_config or "handler" in tool_config:
```
Comment (mpangrazzi) [inline integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py:56]: Edge case: what if someone use `source` or `handler` as keys in state? May worth documenting?

File: integrations/mcp/tests/test_mcp_toolset.py [hunk 9]
```diff
@@ -233,6 +268,121 @@ async def test_toolset_tool_not_found(self):
...
+            "add": {"first_number": "a"},
+            "subtract": {"first_number": "a", "second_number": "b"},
+        }
+        assert toolset_dict["data"]["outputs_to_state"] is not None
+        assert toolset_dict["data"]["outputs_to_string"] is not None
+        # Handler should be serialized as a string
+        assert isinstance(toolset_dict["data"]["outputs_to_string"]["add"]["handler"], str)
+
+        # Test deserialization
+        with patch("haystack_integrations.tools.mcp.mcp_toolset.MCPToolset.__init__", return_value=None) as mock_init:
```
Comment (mpangrazzi) [inline integrations/mcp/tests/test_mcp_toolset.py]: I may miss something, but why not doing a full roundtrip of `to_dict()` / `from_dict()` here (so without mocking)?

**Review by mpangrazzi — COMMENTED** [review]
Looks already good! I've added a few comments. +1 on adding an integration test as @sjrl mentioned.

**Review by sjrl — APPROVED** [review]
Looks good!
